lib/model/internal_link.dart

```diff
@@ -0,0 +1,285 @@
+// it says where to look in a list of messages, but it doesn't filter the list down.
+// In fact, from a brief look at server code, it seems to be *ignored*
+// if you include it in the `narrow` param in get-messages requests.
+// When you want to point the server to a location in a message list, you
+// you do so by passing the `anchor` param.
+Uri narrowLink(PerAccountStore store, Narrow narrow, {int? nearMessageId}) {
+  final apiNarrow = narrow.apiEncode();
+  final fragment = StringBuffer('narrow');
+  for (ApiNarrowElement element in apiNarrow) {
+    fragment.write('/');
+    if (element.negated) {
+      fragment.write('-');
+    }
+
+    if (element is ApiNarrowDm) {
+      final supportsOperatorDm = store.connection.zulipFeatureLevel! >= 177; // TODO(server-7)
+      element = element.resolve(legacy: !supportsOperatorDm);
+    }
+
+    fragment.write('${element.operator}/');
+
+    switch (element) {
+      case ApiNarrowStream():
+        final streamId = element.operand;
+        final name = store.streams[streamId]?.name ?? 'unknown';
+        final slugifiedName = _encodeHashComponent(name.replaceAll(' ', '-'));
+        fragment.write('$streamId-$slugifiedName');
+      case ApiNarrowTopic():
+        fragment.write(_encodeHashComponent(element.operand));
+      case ApiNarrowDmModern():
+        final suffix = element.operand.length >= 3 ? 'group' : 'dm';
+        fragment.write('${element.operand.join(',')}-$suffix');
+      case ApiNarrowPmWith():
+        final suffix = element.operand.length >= 3 ? 'group' : 'pm';
+        fragment.write('${element.operand.join(',')}-$suffix');
+      case ApiNarrowDm():
+        assert(false, 'ApiNarrowDm should have been resolved');
+      case ApiNarrowMessageId():
+        fragment.write(element.operand.toString());
+    }
+  }
+
+  if (nearMessageId != null) {
+    fragment.write('/near/$nearMessageId');
+  }
+
+  return store.account.realmUrl.replace(fragment: fragment.toString());
+}
+
+/// Create a new `Uri` object in relation to a given realmUrl.
+///
+/// Returns `null` if `urlString` could not be parsed as a `Uri`.
+Uri? tryResolveToRealmUrl(String urlString, Uri realmUrl) {
+  try {
+    return realmUrl.resolve(urlString);
+  } on FormatException {
+    return null;
+  }
+}
+
+/// A [Narrow] from a given URL, on `store`'s realm.
+///
+/// Returns `null` if any of the operator/operand pairs are invalid.
+///
+/// Since narrow links can combine operators in ways our [Narrow] type can't
+/// represent, this can also return null for valid narrow links.
+///
+/// This can also return null for some valid narrow links that our Narrow
+/// type *could* accurately represent. We should try to understand these
+/// better, but some kinds will be rare, even unheard-of:
+///   #narrow/stream/1-announce/stream/1-announce (duplicated operator)
+Narrow? parseInternalLink(Uri referenceUrl, PerAccountStore store) {
+  final url = store.account.realmUrl.resolveUri(referenceUrl);
```

Comment:
Hmm interesting. Why the additional `resolveUri` call here?

This makes a significant difference in what the meaning of this function is. It definitely needs to be clearly reflected in the function's doc.

But also in the way this function gets called, it seems redundant — we already pass it a URL that's been resolved to an absolute URL.

It looks like there are no tests exercising this behavior, either — this diff passes all the tests:
```diff
 Narrow? parseInternalLink(Uri referenceUrl, PerAccountStore store) {
-  final url = store.account.realmUrl.resolveUri(referenceUrl);
+  final url = referenceUrl;
```
So if we were to have this behavior, we would definitely want to have tests for it.

Reply:
There was an oversight on my part, it should not be called again. Along with a doc change in https://github.com/zulip/zulip-flutter/pull/305#discussion_r1347181062 this line will be removed.